python/cugraph/structure/hypergraph.py

```diff
@@ -0,0 +1,503 @@
+#     * Neither the name of the Graphistry, Inc nor the
+#       names of its contributors may be used to endorse or promote products
+#       derived from this software without specific prior written permission.
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL Graphistry, Inc BE LIABLE FOR ANY DIRECT,
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+# THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import cudf
+import numpy as np
+from cugraph.structure.graph import Graph
+
+
+def hypergraph(
+    values,
+    columns=None,
+    dropna=True,
+    direct=False,
+    graph_class=Graph,
+    categories=dict(),
+    drop_edge_attrs=False,
+    categorical_metadata=True,
+    SKIP=None,
+    EDGES=None,
+    DELIM="::",
+    SOURCE="src",
+    TARGET="dst",
+    WEIGHTS=None,
+    NODEID="node_id",
+    EVENTID="event_id",
+    ATTRIBID="attrib_id",
+    CATEGORY="category",
+    NODETYPE="node_type",
+    EDGETYPE="edge_type",
+):
+    """
+    Creates a hypergraph out of the given dataframe, returning the graph
+    components as dataframes. The transform reveals relationships between the
+    rows and unique values. This transform is useful for lists of events,
+    samples, relationships, and other structured high-dimensional data.
+
+    The transform creates a node for every row, and turns a row's column
+    entries into node attributes. If direct=False (default), every unique
+    value within a column is also turned into a node. Edges are added to
+    connect a row's nodes to each of its column nodes, or if direct=True, to
+    one another. Nodes are given the attribute specified by ``NODETYPE``
+    that corresponds to the originating column name, or if a row ``EVENTID``.
+
+    Consider a list of events. Each row represents a distinct event, and each
+    column some metadata about an event. If multiple events have common
+    metadata, they will be transitively connected through those metadata
+    values. Conversely, if an event has unique metadata, the unique metadata
+    will turn into nodes that only have connections to the event node.
+
+    For best results, set ``EVENTID`` to a row's unique ID, ``SKIP`` to all
+    non-categorical columns (or ``columns`` to all categorical columns),
+    and ``categories`` to group columns with the same kinds of values.
+
+    Parameters
+    ----------
+    values : cudf.DataFrame
```

Comment:
Just curious, how hard would it be to also accept a distributed dask cudf DataFrame?

Reply:
I haven't tried it, but theoretically we should be able to support it with some testing.